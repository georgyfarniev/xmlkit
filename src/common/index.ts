export * from './types'
export * from './stack'
export * from './serializer'
