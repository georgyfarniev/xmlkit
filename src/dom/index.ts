export * from './xml_cdata'
export * from './xml_comment'
export * from './xml_document'
export * from './xml_element'
export * from './xml_node'
