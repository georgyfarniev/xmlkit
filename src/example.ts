import { XmlDocument } from './xml_document'

const XML_TEXT = `
<?xml version="1.0" encoding="UTF-8"?>
<Root>
  <Child1 foo="bar">Text</Child1>
  <Child2>Text</Child2>
  <Child2>
    <!-- Comment2! -->
    <Grandchild1 fiz="buz" foo="123">Text3</Grandchild1>
    <Grandchild2>Text4</Grandchild2>
  </Child2>
  <Child3/>
  <Child4>
    <LastChild>Text</LastChild>
  </Child4>
  <!-- Comment! -->
</Root>`

async function main() {
  const doc = new XmlDocument(XML_TEXT)

  for (const child of doc.root) {
    console.log(child.name)

    for (const subchild of child) {
      console.log(`\t${subchild.name}`)
    }
  }
}

if (require.main === module) {
  main()
}
