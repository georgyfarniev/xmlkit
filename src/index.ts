
export * from './dom'
export * from './xml_sax'
export * from './xml_stream'
