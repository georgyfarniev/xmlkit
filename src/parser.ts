import sax from 'sax'
import type { Tag } from 'sax';
import {
  IXmlCData,
  IXmlCloseTag,
  IXmlComment,
  IXmlOpenTag,
  IXmlText,
  XmlToken,
  XmlTokenType,
  BaseParser
} from './types';


// Abstraction class to hide sax module behind and allow easy replacement
export class XmlSaxParser implements BaseParser {
  private tokens: XmlToken[] = [];

  constructor(private parser = sax.parser(true)) {
    parser.onopentag = this.beginElement
    parser.onclosetag = this.endElement
    parser.ontext = this.onText
    parser.oncomment = this.onComment
    parser.oncdata = this.onCDATA
  }
 
  //#region  handlers
  private beginElement = (tag: Tag) => {
    const { name, attributes, isSelfClosing } = tag;
    const tok: IXmlOpenTag = {
      name, 
      attrs: attributes,
      selfClosing: isSelfClosing,
      type: XmlTokenType.ElementOpen
    };

    this.tokens.push(tok);
  }

  private endElement = (name: string) => {
    this.tokens.push({
      name,
      type: XmlTokenType.ElementClose
    } as IXmlCloseTag);
  }

  private onText = (text: string) => {
    this.tokens.push({
      text,
      type: XmlTokenType.Text
    } as IXmlText);
  }

  private onCDATA = (text: string) => {
    this.tokens.push({
      text,
      type: XmlTokenType.CDATA
    } as IXmlCData);
  }

  private onComment = (text: string) => {
    this.tokens.push({
      text,
      type: XmlTokenType.Comment
    } as IXmlComment);
  }
  //#endregion

  public getTokens(chunk: string): XmlToken[] {
    this.tokens = []
    this.parser.write(chunk)
    return this.tokens;
  }
}
