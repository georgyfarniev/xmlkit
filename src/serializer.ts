import { Attrs } from './types';

export class XmlSerializer {
  public static openTag(
    name: string,
    attrs: Attrs = {},
    selfClosing = false
  ): string {
    const aStr = Object.entries(attrs || [])
      .reduce((acc, [k, v]) => `${acc}${k}="${v}" `, ' ')
      .trimRight();

    const sc = selfClosing ? '/' : '';
    return `<${name}${aStr}${sc}>`;
  }

  public static closeTag(name: string): string {
    return `</${name}>`;
  }

  public static text(text: string): string {
    return text;
  }

  public static comment(text: string): string {
    return `<!-- ${text} -->`
  }

  public static CDATA(text: string): string {
    return `<![CDATA[${text}]]>`
  }
}