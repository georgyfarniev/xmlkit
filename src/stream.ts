import { Transform, TransformCallback } from 'stream'
import { XmlSaxParser } from './parser';
import { Stack } from './stack'
import {
  XmlTokenType,
  XmlToken,
  IXmlOpenTag,
  IXmlCloseTag,
  IXmlCData,
  IXmlComment,
  IXmlText
} from './types'

export class XmlStream extends Transform {
  private chunk = '';
  private readonly sax: XmlSaxParser = new XmlSaxParser()
  private readonly stack = new Stack<string>();

  constructor(private readonly query: string) {
    super({ objectMode: true })
  }

  public _transform(xml: string, _: string, cb: TransformCallback): void {
    const tokens = this.sax.getTokens(xml);
    const chunks = this.tokensToChunks(tokens, this.query);

    for (const chunk of chunks) this.push(chunk);
    cb();
  }

  private getOpenTagText(tag: IXmlOpenTag) {
    const attrs = Object.entries(tag.attrs || [])
      .reduce((acc, [k, v]) => `${acc}${k}="${v}" `, ' ')
      .trimRight();

    const sc = tag.selfClosing ? '/' : '';
    return `<${tag.name}${attrs}${sc}>`;
  }

  private get path() {
    return this.stack.toArray().join('.');
  }

  private tokensToChunks(tokens: XmlToken[], query) {
    const buf: string[] = [];

    for (const token of tokens) {
      switch (token.type) {
        case XmlTokenType.ElementOpen: {
          const { name } = token as IXmlOpenTag;

          this.stack.push(name);

          const path = this.path

          if (path.startsWith(query)) {
            this.chunk += this.getOpenTagText(token as IXmlOpenTag);
          }

          break;
        }
        case XmlTokenType.ElementClose: {
          const { name } = token as IXmlCloseTag;
          const path = this.path;

          this.chunk += `</${name}>`

          if (path === query) {
            buf.push(this.chunk)
            this.chunk = ''
          }

          this.stack.pop()
          break;
        }
        case XmlTokenType.Text: {
          const { text } = token as IXmlText;
          this.chunk += text
          break;
        }
        case XmlTokenType.Comment: {
          const { text } = token as IXmlComment;
          this.chunk +=  `<!-- ${text} -->`;
          break;
        }
        case XmlTokenType.CDATA: {
          const { text } = token as IXmlCData;
          this.chunk += `<![CDATA[${text}]]>`;
          break;
        }
      }
    }

    return buf;
  }
}
