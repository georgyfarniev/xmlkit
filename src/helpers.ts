import { Readable } from 'stream'

/**
 * Class to generate heavy xml payload for testing purposes.
 * Each value inside contains incremented value for ease of testing
 */
export class TestStream extends Readable {
  private index = 0

  constructor(private readonly total: number = 1 * 1000 * 1000) {
    super()
  }

  public _read(): void {
    const { index, total } = this

    if (index > total) {
      this.push(null)
      return
    }

    const buf = [
      index ? '' : '<root>',
      `<item foo="bar${index}">`,
      `<node1_${index}>text${index}`,
      `<subnode${index}>subtext${index}</subnode${index}>`,
      `</node1_${index}>`,
      '</item>',
      index === total ? '</root>' : ''
    ].join('\n')

    this.push(buf)
    this.index++
  }
}
