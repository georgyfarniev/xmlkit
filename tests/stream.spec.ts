import fs from 'node:fs'
import path from 'node:path'
import { XmlStream } from '../src/stream';

const readResult = async (file: string, ps: string) => {
  const input = fs.createReadStream(
    path.join(__dirname, file)
  );

  const parser = new XmlStream(ps);
  const xml = input.pipe(parser);

  let result = '';
  for await (const chunk of xml) {
    result += chunk;
  }

  return result;
}

describe('XmlStream', () => {
  test('complex example including self-closing elements', async () => {
    await expect(readResult('example.xml', 'root.example.item')).resolves.toMatchSnapshot()
  });
});
