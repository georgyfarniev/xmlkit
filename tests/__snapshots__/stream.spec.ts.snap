// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XmlStream complex example including self-closing elements 1`] = `
"<item>
          <key>value one</key>
          <empty></empty>
          <self_closing/>
      </item><item>
          <key>value two</key>
          <empty></empty>
          <self_closing/>
      </item><item foo="bar">
        <key>
          value two
        </key>
        <text>
        </text>
        <!--  Some comment  -->
        <empty fizz="1"  buzz="2"></empty>
        <![CDATA[<sender>John Smith</sender>]]>
        <self_closing tar="mar"/>
    </item><item>
      <key>
        value two
      </key>
    </item><item></item><item foo="bar"/><item/>"
`;

exports[`XmlStream elements unrelated to selector path must not be present 1`] = `
"<item>
      <key>value one</key>
      <empty></empty>
      <self_closing/>
  </item><item>
      <key>value two</key>
      <empty></empty>
      <self_closing/>
  </item>"
`;
